### PR TITLE
Display Test Failures & Refactor Formpage

### PR DESCRIPTION
**Pull request by HassanAbouelela**
This PR includes a few things:
- An update to the error message component (see 68545056). This is mostly standalone, but it's useful for another change
- Unittest failures now display the reason of failure. This was included in this PR, as the restructuring of formpage made it easier to accomplish
- Refactoring of formpage. This is where the bulk of the changes are, but it's mostly just a straight copy-paste of old code into new files. It does contain a few changes - which should hopefully improve things - such as:
  - Errors are no longer suppressed on submission. We at least alert the user about them. I'd like to expand on this later
  - Submission has been broken up into about a million better-structured functions. This should make it easier to store state and control it in general

Note: I've tagged this as critical, since it dips it's toes into about half the codebase, and other PRs would likely create painful conflicts 😄. Also CJ is active now, and this would be nice to have.

### CODE REVIEW

File: src/api/question.ts [hunk 1]
```diff
@@ -16,3 +16,16 @@ export interface Question {
     data: { [key: string]: string | string[] },
     required: boolean
 }
+
+type unittestError = {
```
Comment (dawnofmidnight) [inline src/api/question.ts:20]: I *think* a common naming convention is to use PascalCase for types, and it'd be consistent with the below interface, but probably not a huge deal, or maybe there's some reasoning I don't understand.

File: src/components/ErrorMessage.tsx [hunk 2]
```diff
@@ -1,10 +1,12 @@
 /** @jsx jsx */
 import { jsx, css } from "@emotion/react";
 import colors from "../colors";
+import {selectable} from "../commonStyles";
```
Comment (dawnofmidnight) [inline src/components/ErrorMessage.tsx:4]: The braces here are inconsistent with the ones two lines above, but that's a style thing that doesn't really matter I don't think. It does seem to be consistent with a lot of the rest of the codebase (or at least, what I can see in this diff).

**Review by dawnofmidnight — DISMISSED** [review]

File: src/components/ErrorMessage.tsx (hunk 2 above)
Comment (HassanAbouelela) [reply src/components/ErrorMessage.tsx:4]: Yeah that sounds like a fine change. acb7df25

File: src/api/question.ts (hunk 1 above)
Comment (HassanAbouelela) [reply src/api/question.ts]: acb7df25

File: src/pages/FormPage/ErrorPage.tsx [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+import {Form} from "../../api/forms";
+import {selectable, submitStyles, unselectable} from "../../commonStyles";
+
+import Navigation from "./Navigation";
+
+
+interface ErrorProps {
+    form: Form
+    questions: RenderedQuestion[]
```
Comment (MarkKoz) [inline src/pages/FormPage/ErrorPage.tsx:17]: Why is this included if it's never used?

File: src/components/ErrorMessage.tsx [hunk 4]
```diff
@@ -1,10 +1,12 @@
...
-import { jsx, css } from "@emotion/react";
+import {jsx, css} from "@emotion/react";
 import colors from "../colors";
+import {selectable} from "../commonStyles";
 
 interface ErrorMessageProps {
     show: boolean,
-    message: string
+    message: string,
+    innerElement?: JSX.Element,
```
Comment (MarkKoz) [inline src/components/ErrorMessage.tsx:9]: `message` and `innerElement` seem mutually exclusive, so this seems wrong. Not sure what a better design would be.

File: src/pages/FormPage/submit.ts [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+            .catch(error => {
+                if (!error.response) {
+                    throw error;
+                }
+
+                switch (error.response.status) {
+                    case 422:
+                        // TODO: Re-enable this once we have better state management
+                        // setState(FormState.VALIDATION_ERROR);
+                        showUnitTestFailures(refMap, error.response.data);
```
Comment (MarkKoz) [inline src/pages/FormPage/submit.ts:73]: Is the idea that in the future, this function should just set a state, and something consuming that state will be responsible for doing `showUnitTestFailures`?

File: src/components/ErrorMessage.tsx (hunk 4 above)
Comment (HassanAbouelela) [reply src/components/ErrorMessage.tsx:9]: There's always the option to combine them into one ORed field, and then add instance checks in the code. I can do that.

File: src/components/Question.tsx [hunk 6]
```diff
@@ -299,18 +315,23 @@ class RenderedQuestion extends React.Component<QuestionProp> {
...
-            let error = "";
-            if (typeof rawError === "string") {
-                error = rawError;
+
+            let error;
+            if (this.props.question.type === QuestionType.Code && this.props.public_state.get("unittestsFailed")) {
+                error = this.generateUnitTestErrorMessage(valid);
+            } else {
+                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+                // @ts-ignore
```
[the remark this answers is not in the document]
Comment (HassanAbouelela) [reply src/components/Question.tsx:324]: It's always a string right now, but there's no way to legally assert that unless you add a type check.

Since we're setting the error variable based on this, another check will need to be added to handle the case where the error variable was not set. This never happens though, so it would be a bunch of fun error handling for an impossibility.

The larger issue here is that state is a map with no hard defined rules on type. The same is true of many things around the codebase, such as setting the form ID, which has to be asserted to being a string.

One thing that could alleviate this would be defining an interface for state, but even then you're going to have to handle a few null checks which we know are impossible right now.

File: src/pages/FormPage/ErrorPage.tsx (hunk 3 above)
Comment (HassanAbouelela) [reply src/pages/FormPage/ErrorPage.tsx:17]: I initially included some serialization and deserialization for the refresh button, and just didn't get around to removing it. I'll do that

File: src/pages/FormPage/submit.ts (hunk 5 above)
Comment (HassanAbouelela) [reply src/pages/FormPage/submit.ts:73]: That's one option, but mainly this is going to be important once we're switching between the sending state, and some other state. *Some* state change is going to be necessary.

Validation error can be merged into the initial state, but I set it to its own type here in case it proves useful for something else (such as moving the validation logic, or rewriting the scroll logic).

File: src/components/ErrorMessage.tsx (hunk 4 above)
Comment (HassanAbouelela) [reply src/components/ErrorMessage.tsx]: 9696f526

File: src/pages/FormPage/ErrorPage.tsx (hunk 3 above)
Comment (HassanAbouelela) [reply src/pages/FormPage/ErrorPage.tsx:17]: b83dff71

File: src/components/Question.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (HassanAbouelela) [reply src/components/Question.tsx]: This took a lot more work than expected, but it's done now. 09f25bed

**Review by MarkKoz — APPROVED** [review]
